Build failure: tftp-hpa

## Steps To Reproduce

1. Install tftp-hpa via whichever mechanism.

Almost assuredly another victim of the gcc 14 migration.

## Build log

```
error: builder for '/nix/store/99zixhn7xs9cc4gq99jnlqrqpk8z0yci-tftp-hpa-5.2.drv' failed with exit code 2;
       last 25 log lines:
       > gcc -g -O2 -W -Wall -Wpointer-arith -Wbad-function-cast -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -Winline -Wwrite-strings -Wundef -Wshadow -Wsign-compare -pipe -fno-strict-aliasing -I/build/tftp-hpa-5.2 -c xmalloc.c
       > gcc -g -O2 -W -Wall -Wpointer-arith -Wbad-function-cast -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -Winline -Wwrite-strings -Wundef -Wshadow -Wsign-compare -pipe -fno-strict-aliasing -I/build/tftp-hpa-5.2 -c xstrdup.c
       > rm -f libxtra.a
       > ar cq libxtra.a xmalloc.o xstrdup.o
       > ranlib libxtra.a
       > make[1]: Leaving directory '/build/tftp-hpa-5.2/lib'
       > make -C  common
       > make[1]: Entering directory '/build/tftp-hpa-5.2/common'
       > gcc -g -O2 -W -Wall -Wpointer-arith -Wbad-function-cast -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -Winline -Wwrite-strings -Wundef -Wshadow -Wsign-compare -pipe -fno-strict-aliasing -I/build/tftp-hpa-5.2 -c tftpsubs.c
       > rm -f libcommon.a
       > ar cq libcommon.a tftpsubs.o
       > ranlib libcommon.a
       > make[1]: Leaving directory '/build/tftp-hpa-5.2/common'
       > make -C  tftp
       > make[1]: Entering directory '/build/tftp-hpa-5.2/tftp'
       > gcc -g -O2 -W -Wall -Wpointer-arith -Wbad-function-cast -Wstrict-prototypes -Wmissing-prototypes -Wmissing-declarations -Wnested-externs -Winline -Wwrite-strings -Wundef -Wshadow -Wsign-compare -pipe -fno-strict-aliasing -I/build/tftp-hpa-5.2 -c tftp.c
       > tftp.c: In function 'tftp_sendfile':
       > tftp.c:88:5: error: implicit declaration of function 'bsd_signal'; did you mean 'ssignal'? [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration8;;]
       >    88 |     bsd_signal(SIGALRM, timer);
       >       |     ^~~~~~~~~~
       >       |     ssignal
       > tftp.c:88:5: warning: nested extern declaration of 'bsd_signal' [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wnested-externs-Wnested-externs8;;]
       > make[1]: *** [../MRULES:6: tftp.o] Error 1
       > make[1]: Leaving directory '/build/tftp-hpa-5.2/tftp'
       > make: *** [Makefile:7: tftp.build] Error 2
       For full logs, run 'nix log /nix/store/99zixhn7xs9cc4gq99jnlqrqpk8z0yci-tftp-hpa-5.2.drv'.
```

Full build failure log at https://gist.github.com/nipsy/92bfb490a8f754b4257b8b5421be0cbe.

## Metadata

```
# nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.6, NixOS, 25.05 (Warbler), 25.05.20241219.d70bd19`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.11`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/8fwsiv0hd7nw1brkvka0jf1frk3m7qkr-source`
```

## Notify maintainers

@7c6f434c

looks like using the git source fixes. also seems to be used by Debian: https://packages.debian.org/sid/source/tftp-hpa

<details><summary>use git src</summary>

```diff
diff --git a/pkgs/by-name/tf/tftp-hpa/package.nix b/pkgs/by-name/tf/tftp-hpa/package.nix
index f9946be7423a..9bd3a3f33fb9 100644
--- a/pkgs/by-name/tf/tftp-hpa/package.nix
+++ b/pkgs/by-name/tf/tftp-hpa/package.nix
@@ -1,22 +1,24 @@
 {
   lib,
   stdenv,
-  fetchurl,
+  fetchzip,
+  autoconf,
+  automake,
 }:
 
 stdenv.mkDerivation rec {
   pname = "tftp-hpa";
-  version = "5.2";
-  src = fetchurl {
-    url = "mirror://kernel/software/network/tftp/tftp-hpa/${pname}-${version}.tar.xz";
-    sha256 = "12vidchglhyc20znq5wdsbhi9mqg90jnl7qr9qs8hbvaz4fkdvmg";
+  version = "5.2-unstable-2024-06-10";
+  src = fetchzip {
+    url = "https://git.kernel.org/pub/scm/network/tftp/tftp-hpa.git/snapshot/tftp-hpa-2c86ff58dcc003107b47f2d35aa0fdc4a3fd95e1.tar.gz";
+    hash = "sha256-lTMldYO/cZdLj0UjOPPBHfYf2GBG0O+5lhP9ikqn3tY=";
   };
 
-  # Workaround build failure on -fno-common toolchains like upstream
-  # gcc-10. Otherwise build fails as:
-  #   ld: main.o:/build/tftp-hpa-5.2/tftp/main.c:98: multiple definition of
-  #     `toplevel'; tftp.o:/build/tftp-hpa-5.2/tftp/tftp.c:51: first defined here
-  env.NIX_CFLAGS_COMPILE = "-fcommon";
+  nativeBuildInputs = [ autoconf automake ];
+
+  preConfigure = ''
+    make configure
+  '';
 
   meta = with lib; {
     description = "TFTP tools - a lot of fixes on top of BSD TFTP";

```
</details>

--- a/pkgs/applications/video/mplayer/default.nix
+++ b/pkgs/applications/video/mplayer/default.nix
@@ -120,12 +120,12 @@ in
 
 stdenv.mkDerivation rec {
   pname = "mplayer";
-  version = "1.5-unstable-2024-07-03";
+  version = "1.5-unstable-2024-12-21";
 
   src = fetchsvn {
     url = "svn://svn.mplayerhq.hu/mplayer/trunk";
-    rev = "38637";
-    hash = "sha256-9KQOB6QIs1VZhazJqW8dY4ASiMgoxV6davfpKgLPbmE=";
+    rev = "38668";
+    hash = "sha256-ezWYBkhiSBgf/SeTrO6sKGbL/IrX+82KXCIlqYMEtgY=";
   };
 
   prePatch = ''
@@ -276,7 +276,8 @@ stdenv.mkDerivation rec {
     description = "Movie player that supports many video formats";
     homepage = "http://mplayerhq.hu";
     license = licenses.gpl2Only;
-    maintainers = [ ];
+    # Picking it up: no idea about the origin of some choices (but seems fine)
+    maintainers = [ maintainers.raskin ];
     platforms = [
       "i686-linux"
       "x86_64-linux"

--- a/pkgs/by-name/tf/tftp-hpa/package.nix
+++ b/pkgs/by-name/tf/tftp-hpa/package.nix
@@ -1,15 +1,18 @@
 {
   lib,
   stdenv,
-  fetchurl,
+  fetchgit,
+  autoconf,
+  automake,
 }:
 
 stdenv.mkDerivation rec {
   pname = "tftp-hpa";
-  version = "5.2";
-  src = fetchurl {
-    url = "mirror://kernel/software/network/tftp/tftp-hpa/${pname}-${version}.tar.xz";
-    sha256 = "12vidchglhyc20znq5wdsbhi9mqg90jnl7qr9qs8hbvaz4fkdvmg";
+  version = "5.2-untagged-2024-06-10";
+  src = fetchgit {
+    url = "git://git.kernel.org/pub/scm/network/tftp/tftp-hpa.git";
+    hash = "sha256-lTMldYO/cZdLj0UjOPPBHfYf2GBG0O+5lhP9ikqn3tY=";
+    rev = "2c86ff58dcc003107b47f2d35aa0fdc4a3fd95e1";
   };
 
   # Workaround build failure on -fno-common toolchains like upstream
@@ -17,6 +20,15 @@ stdenv.mkDerivation rec {
   #   ld: main.o:/build/tftp-hpa-5.2/tftp/main.c:98: multiple definition of
   #     `toplevel'; tftp.o:/build/tftp-hpa-5.2/tftp/tftp.c:51: first defined here
   env.NIX_CFLAGS_COMPILE = "-fcommon";
+
+  preConfigure = ''
+    ./autogen.sh
+  '';
+
+  nativeBuildInputs = [
+    autoconf
+    automake
+  ];
 
   meta = with lib; {
     description = "TFTP tools - a lot of fixes on top of BSD TFTP";

--- a/pkgs/games/xpilot/default.nix
+++ b/pkgs/games/xpilot/default.nix
@@ -32,6 +32,9 @@ stdenv.mkDerivation rec {
     zlib
     libXxf86misc
   ];
+
+  patches = [ ./xpilot-ng-gcc-14-fix.patch ];
+
   meta = with lib; {
     description = "Multiplayer X11 space combat game";
     homepage = "http://xpilot.sf.net/";

--- a/pkgs/games/xpilot/xpilot-ng-gcc-14-fix.patch
+++ b/pkgs/games/xpilot/xpilot-ng-gcc-14-fix.patch
@@ -1,0 +1,168 @@
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/mapedit/proto.h xpilot-ng-4.7.3-fixed/src/mapedit/proto.h
+--- xpilot-ng-4.7.3/src/mapedit/proto.h	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/mapedit/proto.h	2024-12-31 14:25:55.790083767 +0100
+@@ -73,15 +73,15 @@
+ int ResizeWidth(HandlerInfo_t info);
+ int ResizeHeight(HandlerInfo_t info);
+ int OpenPreferencesPopup(HandlerInfo_t info);
+-int OpenMapInfoPopup(void);
+-int OpenRobotsPopup(void);
+-int OpenVisibilityPopup(void);
+-int OpenCannonsPopup(void);
+-int OpenRoundsPopup(void);
+-int OpenInitItemsPopup(void);
+-int OpenMaxItemsPopup(void);
+-int OpenProbsPopup(void);
+-int OpenScoringPopup(void);
++int OpenMapInfoPopup(HandlerInfo_t info);
++int OpenRobotsPopup(HandlerInfo_t info);
++int OpenVisibilityPopup(HandlerInfo_t info);
++int OpenCannonsPopup(HandlerInfo_t info);
++int OpenRoundsPopup(HandlerInfo_t info);
++int OpenInitItemsPopup(HandlerInfo_t info);
++int OpenMaxItemsPopup(HandlerInfo_t info);
++int OpenProbsPopup(HandlerInfo_t info);
++int OpenScoringPopup(HandlerInfo_t info);
+ int ValidateCoordHandler(HandlerInfo_t info);
+ int ShowHoles(HandlerInfo_t info);
+ char MapData(int x, int y);
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/mapedit/tools.c xpilot-ng-4.7.3-fixed/src/mapedit/tools.c
+--- xpilot-ng-4.7.3/src/mapedit/tools.c	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/mapedit/tools.c	2024-12-31 14:26:29.168082279 +0100
+@@ -747,7 +747,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenMapInfoPopup()
++int OpenMapInfoPopup(HandlerInfo_t info)
+ {
+     Window *temp;
+ 
+@@ -790,7 +790,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenRobotsPopup()
++int OpenRobotsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, robots);
+     return 0;
+@@ -801,7 +801,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenVisibilityPopup()
++int OpenVisibilityPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, visibility);
+     return 0;
+@@ -812,7 +812,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenCannonsPopup()
++int OpenCannonsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, cannons);
+     return 0;
+@@ -823,7 +823,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenRoundsPopup()
++int OpenRoundsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, rounds);
+     return 0;
+@@ -834,7 +834,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenInitItemsPopup()
++int OpenInitItemsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, inititems);
+     return 0;
+@@ -845,7 +845,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenMaxItemsPopup()
++int OpenMaxItemsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, maxitems);
+     return 0;
+@@ -856,7 +856,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenProbsPopup()
++int OpenProbsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, probs);
+     return 0;
+@@ -867,7 +867,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenScoringPopup()
++int OpenScoringPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, scoring);
+     return 0;
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/server/suibotdef.c xpilot-ng-4.7.3-fixed/src/server/suibotdef.c
+--- xpilot-ng-4.7.3/src/server/suibotdef.c	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/server/suibotdef.c	2024-12-31 14:07:28.316133143 +0100
+@@ -338,12 +338,6 @@
+ 
+ /* } */
+ 
+-struct collans {
+-    int line;
+-    int point;
+-    clvec_t moved;
+-};
+-
+ static bool Wall_in_between_points(int cx1, int cy1, int cx2, int cy2);
+ static bool Wall_in_between_points(int cx1, int cy1, int cx2, int cy2){ /* Wall between two given points?*/
+   
+@@ -405,7 +399,7 @@
+   /* ignore if there is enough time to deal with this object  later */
+   if((time_until_closest < 0) || (time_until_closest > maxtime))
+     /*option instead of fixed value: options.dodgetime))*/
+-    return;
++    return false;
+ 
+   /* get the square of the distance */
+   sqdistance =
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/server/walls.c xpilot-ng-4.7.3-fixed/src/server/walls.c
+--- xpilot-ng-4.7.3/src/server/walls.c	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/server/walls.c	2024-12-31 14:09:46.086127001 +0100
+@@ -54,12 +54,6 @@
+ #error "This code assumes that negative numbers round upwards."
+ #endif
+ 
+-struct collans {
+-    int line;
+-    int point;
+-    clvec_t moved;
+-};
+-
+ struct tl2 {
+     int base;
+     int x;
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/server/walls.h xpilot-ng-4.7.3-fixed/src/server/walls.h
+--- xpilot-ng-4.7.3/src/server/walls.h	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/server/walls.h	2024-12-31 14:04:50.198140193 +0100
+@@ -187,4 +187,11 @@
+ 
+ extern int num_polys, num_pstyles, num_estyles, num_bstyles;
+ 
++struct collans {
++    int line;
++    int point;
++    clvec_t moved;
++};
++
++extern void Move_point(const move_t *move, struct collans *answer);
+ #endif

